doc/modules/gaussian_process.rst

```diff
@@ -605,6 +605,100 @@ shown in the following figure:
    :target: ../auto_examples/gaussian_process/plot_gpr_prior_posterior.html
    :align: center
 
+Custom kernel
+------------------
+If you want to implement a new custom gaussian process kernel that is scikit-learn-compatible,
+you should subclass `sklearn.gaussian_process.kernels.Kernel` and implement
+the abstract methods `__call__`, `diag`, and `is_stationary`.
+
+Here is a step-by-step construction of the Sigmoid Kernel
+
+.. math::
+   k(x_i, x_j)= \tanh(\alpha x_i \cdot x_j + \sigma)
+
+
+First, we read in :math:`\alpha` and :math:`\sigma` and store these
+hyperparameters::
+
+   from sklearn.gaussian_process.kernels import Kernel, Hyperparameter
+
+   class MySigmoidKernel(Kernel):
+      def __init__(self, alpha_0=0.05,
+                  alpha_0_bounds=(1e-5, 1e5),
+                  sigma_0=5,
+                  sigma_0_bounds=(1e-5, 1e5)):
+         self.alpha_0 = alpha_0
+         self.alpha_0_bounds = alpha_0_bounds
+         self.sigma_0 = sigma_0
+         self.sigma_0_bounds = sigma_0_bounds
+
+      @property
+      def hyperparameter_scaling(self):
+         return Hyperparameter("alpha_0", "numeric", self.alpha_0_bounds)
+
+      @property
+      def hyperparameter_shifting(self):
+         return Hyperparameter("sigma_0", "numeric", self.sigma_0_bounds)
+
+Note that your custom kernel must be positive-definite for the choice of the
+hyperparameters.
+
+Now, from the kernel formula, we know that this kernel is non-stationary. If your custom
+kernel is stationary, you could subclass
+`sklearn.gaussian_process.kernels.StationaryKernelMixin` and skip this step; otherwise, we
+implement the method `is_stationary` as follows::
+
+      def is_stationary(self):
+         return False
+
+To calculate the diagonal, we will apply the kernel function on `X` (if your custom
+kernel is normalized, i.e. :math:`k(x,x)=1`, then you can subclass
+`sklearn.gaussian_process.kernels.NormalizedKernelMixin` and skip this step)::
+
+      def diag(self, X):
+         return np.apply_along_axis(self, 1, X).ravel()
+
+Finally, for the `__call__` method, we have to calculate the gradient with respect
+to the log-transformed hyperparameters. In this case, we have two hyperparameters:
```

Comment:
I may be mistaken, but from looking at `DotProduct.__call__`, the gradients seems to be with respect to the hyperparameters directly.

Reply:
From https://github.com/scikit-learn/scikit-learn/pull/8373#issuecomment-280844796, I believe the gradient is taken wrt to the log-transformed hyperparameters, which explains why the gradient is `2 \sigma^2` instead of `2 \sigma` :/

Reply:
Thank you for the link, you are correct. It looks like the docstring to `DotProduct.__call__` still needs to be updated to reflect this.